cunumeric/utils.py

```diff
@@ -148,36 +162,61 @@ def matmul_modes(a_ndim, b_ndim):
     return (a_modes, b_modes, out_modes)
 
 
-def tensordot_modes(a_ndim, b_ndim, axes):
-    if isinstance(axes, int):
-        if axes > a_ndim or axes > b_ndim:
+Axis = Sequence[int]
+AxesLikeTuple = Union[
+    Tuple[int, int],
+    Tuple[int, Axis],
+    Tuple[Axis, int],
+    Tuple[Axis, Axis],
+]
+AxesLike = Union[int, AxesLikeTuple]
+
+
+def tensordot_modes(a_ndim: int, b_ndim: int, axes: AxesLike) -> Modes:
+    def convert_int_axes(axes: int) -> tuple[Axis, Axis]:
+        return list(range(a_ndim - axes, a_ndim)), list(range(axes))
+
+    def convert_seq_axes(axes: AxesLikeTuple) -> tuple[Axis, Axis]:
+        a_axis, b_axis = axes
+        return (
+            [a_axis] if isinstance(a_axis, int) else list(a_axis),
+            [b_axis] if isinstance(b_axis, int) else list(b_axis),
+        )
+
+    def convert_axes(axes: AxesLike) -> tuple[Axis, Axis]:
+        if isinstance(axes, int):
+            a_axis, b_axis = convert_int_axes(axes)
+        else:
+            a_axis, b_axis = convert_seq_axes(axes)
+
+        return (
+            [ax + a_ndim if ax < 0 else ax for ax in a_axis],
+            [ax + b_ndim if ax < 0 else ax for ax in b_axis],
+        )
```

Comment:
Just noting that any time there is a complicated set of possible type adaptations (e.g. because various "convenience types" are desirable to handle) splitting things out like this helps to keep the intermediate conversion types organized.